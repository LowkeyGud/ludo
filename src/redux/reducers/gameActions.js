import {
  SafeSpots,
  StarSpots,
  startingPoints,
  turningPoints,
  victoryStart,
} from '../../helpers/PlotData';
import {playSound} from '../../helpers/SoundUtility';
import {selectCurrentPositions, selectDiceNo} from './gameSelector';
import {
  announceWinner,
  disableTouch,
  unfreezeDice,
  updateFireworks,
  updatePlayerChance,
  updatePlayerPieceValue,
} from './gameSlice';

const delay = duration => new Promise(resolve => setTimeout(resolve, duration));

const checkWinningCriteria = pieces => {
  for (let piece of pieces) {
    if (piece.travelCount < 57) {
      return false;
    }
  }
  return true;
};

export const handleForwardThunk = (playerNo, id, pos) => {
  // 3 C1 27

  return async (dispatch, getState) => {
    const state = getState();
    const plottedPieces = selectCurrentPositions(state);
    const diceNo = selectDiceNo(state);

    let alpha =
      playerNo === 1 ? 'A' : playerNo === 2 ? 'B' : playerNo === 3 ? 'C' : 'D';

    const peicesAtPosition = plottedPieces.filter(e => e.pos === pos);
    const piece =
      peicesAtPosition[peicesAtPosition.findIndex(e => e.id[0] == alpha)];

    dispatch(disableTouch());

    let finalPath = piece.pos;
    const beforePlayerPiece = state.game[`player${playerNo}`].find(
      e => e.id === id,
    );
    let travelCount = beforePlayerPiece?.travelCount;

    // for loop starts from 0 to diceNo - 1
    for (let i = 0; i < diceNo; i++) {
      const updatedPosition = getState();
      const playerPiece = updatedPosition.game[`player${playerNo}`].find(
        e => e.id === id,
      );
      let path = playerPiece?.pos + 1;
      if (turningPoints.includes(path) && turningPoints[playerNo - 1] == path) {
        path = victoryStart[playerNo - 1];
      }

      if (path == 53) {
        path = 1;
      }

      finalPath = path;
      travelCount += 1;

      dispatch(
        updatePlayerPieceValue({
          playerNo: `player${playerNo}`,
          pieceId: playerPiece?.id,
          pos: path,
          travelCount: travelCount,
        }),
      );
      playSound('pile_move');
      await delay(100);
    }
    // for loop ends here

    const updatedState = getState();
    const updatedPlottedPieces = selectCurrentPositions(updatedState);
    const finalPlot = updatedPlottedPieces.filter(e => e.pos == finalPath);
    const ids = finalPlot.map(e => e.id[0]);

    const uniqueIDs = new Set(ids);
    const areDifferentIds = uniqueIDs.size > 1;

    console.log('ids', ids);
    console.log('uniqueIDs', uniqueIDs);
    console.log('areDifferentIds', areDifferentIds);

    if (SafeSpots.includes(finalPath) || StarSpots.includes(finalPath)) {
      playSound('safe_spot');
    }

    // if statement starts here
    if (
      areDifferentIds &&
      !SafeSpots.includes(finalPlot[0].pos) &&
      !StarSpots.includes(finalPlot[0].pos)
    ) {
      const enemyPiece = finalPlot.find(p => p.id[0] !== id[0]);
      const enemyId = enemyPiece.id[0];
      let no =
        enemyId === 'A' ? 1 : enemyId === 'B' ? 2 : enemyId === 'C' ? 3 : 4;

      let backwordPath = startingPoints[no - 1];
      let i = enemyPiece.pos;
      playSound('collide');
      // kill the enemy piece with backward movement to their base
      while (i !== backwordPath) {
        dispatch(
          updatePlayerPieceValue({
            playerNo: `player${no}`,
            pieceId: enemyPiece.id,
            pos: i,
            travelCount: 0,
          }),
        );

        await delay(0.001);
        i--;
        if (i === 0) {
          i = 52;
        }
      }

      dispatch(
        updatePlayerPieceValue({
          playerNo: `player${no}`,
          pieceId: enemyPiece.id,
          pos: 0,
          travelCount: 0,
        }),
      );

      dispatch(unfreezeDice());
      return;
    }
    // if statement ends here

    if (diceNo == 6 || travelCount == 57) {
      dispatch(updatePlayerChance({chancePlayer: playerNo}));

      if (travelCount == 57) {
        playSound('home_win');
        const finalPlayerState = getState();
        const playerAllPieces = finalPlayerState.game[`player${playerNo}`];

        if (checkWinningCriteria(playerAllPieces)) {
          dispatch(announceWinner(playerNo));
          playSound('cheer');
          return;
        }

        dispatch(updateFireworks(true));
        dispatch(unfreezeDice());
        return;
      }
    } else {
      let chancePlayer = playerNo + 1;
      if (chancePlayer >= 5) {
        chancePlayer = 1;
      }
      dispatch(updatePlayerChance({chancePlayer}));
    }
  };
};
