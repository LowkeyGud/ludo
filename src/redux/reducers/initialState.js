const player1InitialState = [
  // {id: 'A1', pos: 226, travelCount: 57},
  {id: 'A1', pos: 0, travelCount: 0},
  {id: 'A2', pos: 0, travelCount: 0},
  {id: 'A3', pos: 0, travelCount: 0},
  {id: 'A4', pos: 0, travelCount: 0},
];

const player2InitialState = [
  // {id: 'B1', pos: 336, travelCount: 57},
  {id: 'B1', pos: 0, travelCount: 0},
  {id: 'B2', pos: 0, travelCount: 0},
  {id: 'B3', pos: 0, travelCount: 0},
  {id: 'B4', pos: 0, travelCount: 0},
];

const player3InitialState = [
  // {id: 'C1', pos: 446, travelCount: 57},
  {id: 'C1', pos: 0, travelCount: 0},
  {id: 'C2', pos: 0, travelCount: 0},
  {id: 'C3', pos: 0, travelCount: 0},
  {id: 'C4', pos: 0, travelCount: 0},
];

const player4InitialState = [
  // {id: 'D1', pos: 116, travelCount: 57},
  {id: 'D1', pos: 0, travelCount: 0},
  {id: 'D2', pos: 0, travelCount: 0},
  {id: 'D3', pos: 0, travelCount: 0},
  {id: 'D4', pos: 0, travelCount: 0},
];

const testCurrentPositionsOneHome = [
  {id: 'A1', pos: 226},
  {id: 'B1', pos: 336},
  {id: 'C1', pos: 446},
  {id: 'D1', pos: 116},
];

export const initialState = {
  player1: player1InitialState,
  player2: player2InitialState,
  player3: player3InitialState,
  player4: player4InitialState,
  chancePlayer: 1,
  diceNo: 1,
  isDiceRolled: false,
  pileSelectionPlayer: -1, // no player selected
  cellSelectionPlayer: -1,
  diceTouchBlock: false,
  currentPositions: [],
  // currentPositions: testCurrentPositionsOneHome,
  fireworks: false,
  winner: null,
};
