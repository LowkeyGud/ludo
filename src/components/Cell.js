import {memo, useCallback, useMemo} from 'react';
import {StyleSheet, View} from 'react-native';
import {RFValue} from 'react-native-responsive-fontsize';
import Ionicons from 'react-native-vector-icons/Ionicons';
import {useDispatch, useSelector} from 'react-redux';
import {Colors} from '../constants/Colors';
import {ArrowSpot, SafeSpots, StarSpots} from '../helpers/PlotData';
import {handleForwardThunk} from '../redux/reducers/gameActions';
import {selectCurrentPositions} from '../redux/reducers/gameSelector';
import Pile from './Pile';

const Cell = ({id, color}) => {
  const dispatch = useDispatch();
  const plotedPieces = useSelector(selectCurrentPositions);

  const isSafeSpot = useMemo(() => SafeSpots.includes(id), [id]);
  const isStarSpot = useMemo(() => StarSpots.includes(id), [id]);
  const isArrowSpot = useMemo(() => ArrowSpot.includes(id), [id]);

  const arrowValue =
    id === 38 ? '180deg' : id === 25 ? '90deg' : id === 12 ? '0deg' : '-90deg';

  const peicesAtPosition = useMemo(
    () => plotedPieces.filter(item => item.pos == id),
    [plotedPieces, id],
  );

  const handlePress = useCallback(
    (playerNo, pieceId) => {
      dispatch(handleForwardThunk(playerNo, pieceId, id));
    },
    [dispatch, id],
  );

  return (
    <View
      style={[
        styles.container,
        {backgroundColor: isSafeSpot ? color : Colors.white},
      ]}>
      {isStarSpot && (
        <Ionicons name="star-outline" size={20} color={Colors.grey} />
      )}
      {isArrowSpot && (
        <Ionicons
          name="arrow-forward-outline"
          size={RFValue(12)}
          color={Colors.grey}
          style={{transform: [{rotate: arrowValue}]}}
        />
      )}
      {peicesAtPosition.map((piece, index) => {
        const playerNo =
          piece.id[0] === 'A'
            ? 1
            : piece.id[0] === 'B'
            ? 2
            : piece.id[0] === 'C'
            ? 3
            : 4;
        const pieceColor =
          piece.id[0] === 'A'
            ? Colors.red
            : piece.id[0] === 'B'
            ? Colors.green
            : piece.id[0] === 'C'
            ? Colors.yellow
            : Colors.blue;

        return (
          <View
            key={piece.id}
            style={[
              styles.pieceContainer,
              {
                transform: [
                  {
                    scale: peicesAtPosition.length === 1 ? 1 : 0.7,
                  },
                  {
                    translateX:
                      peicesAtPosition.length === 1
                        ? 0
                        : index % 2 === 0
                        ? -6
                        : 6,
                  },
                  {
                    translateY:
                      peicesAtPosition.length === 1 ? 0 : index < 2 ? -6 : 6,
                  },
                ],
              },
            ]}>
            <Pile
              color={pieceColor}
              player={playerNo}
              cell={true}
              pieceId={piece.id}
              onPress={() => handlePress(playerNo, piece.id)}
            />
          </View>
        );
      })}
    </View>
  );
};

export default memo(Cell);

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
    height: '100%',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: 'grey',
    // zIndex: -109,
  },
  pieceContainer: {
    position: 'absolute',
    top: 0,
    bottom: 0,
  },
});
