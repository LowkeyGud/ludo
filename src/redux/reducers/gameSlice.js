import {createSlice} from '@reduxjs/toolkit';
import {initialState} from './initialState';

export const gameSlice = createSlice({
  name: 'game',
  initialState: initialState,
  reducers: {
    resetGame: () => initialState,
    updateDiceNumber: (state, action) => {
      state.diceNo = action.payload.diceNo;
      state.isDiceRolled = false;
    },
    enablePileSelection: (state, action) => {
      state.diceTouchBlock = true;
      state.pileSelectionPlayer = action.payload.playerNo;
    },
    enableCellSelection: (state, action) => {
      state.diceTouchBlock = true;
      state.cellSelectionPlayer = action.payload.playerNo;
    },
    disableTouch: state => {
      state.diceTouchBlock = true;
      state.pileSelectionPlayer = -1;
      state.cellSelectionPlayer = -1;
    },
    unfreezeDice: state => {
      state.diceTouchBlock = false;
      state.isDiceRolled = false;
    },
    updateFireworks: (state, action) => {
      state.fireworks = action.payload;
    },
    announceWinner: (state, action) => {
      state.winner = action.payload;
    },
    updatePlayerChance: (state, action) => {
      state.chancePlayer = action.payload.chancePlayer;
      state.diceTouchBlock = false;
      state.isDiceRolled = false;
    },
    updatePlayerPieceValue: (state, action) => {
      // Destructure the action payload to extract the necessary data
      const {playerNo, pieceId, pos, travelCount} = action.payload;

      // Access the pieces of the specific player using playerNo
      const playerPieces = state[playerNo];

      // Find the specific piece that matches the given pieceId
      const piece = playerPieces.find(p => p.id === pieceId);

      // Reset pile selection player to -1, indicating no pile is being selected
      state.pileSelectionPlayer = -1;

      // Check if the piece was found (i.e., it exists for that player)
      if (piece) {
        // Update the position and travelCount of the piece
        piece.pos = pos;
        piece.travelCount = travelCount;

        // Find the index of the piece in the currentPositions array
        const currentPositionIdx = state.currentPositions.findIndex(
          p => p.id === pieceId,
        );

        // If the new position is 0 (i.e., the piece is off the board or reset)
        if (pos === 0) {
          // If the piece exists in currentPositions, remove it
          if (currentPositionIdx !== -1) {
            state.currentPositions.splice(currentPositionIdx, 1);
          }
        } else {
          // If the new position is not 0
          if (currentPositionIdx !== -1) {
            // If the piece exists in currentPositions, update its position
            state.currentPositions[currentPositionIdx] = {id: pieceId, pos};
          } else {
            // If the piece is not in currentPositions, add it with the new position
            state.currentPositions.push({id: pieceId, pos});
          }
        }
      }
    },
  },
});

export const {
  resetGame,
  updateDiceNumber,
  enablePileSelection,
  enableCellSelection,
  disableTouch,
  unfreezeDice,
  updateFireworks,
  announceWinner,
  updatePlayerChance,
  updatePlayerPieceValue,
} = gameSlice.actions;

export default gameSlice.reducer;
