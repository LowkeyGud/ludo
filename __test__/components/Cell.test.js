// Cell.test.js
import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';

import * as redux from 'react-redux';
import Cell from '../../src/components/Cell';
import {handleForwardThunk} from '../../src/redux/reducers/gameActions';

// --- Mock External Dependencies ---

// Mock Ionicons so it doesn’t try to load actual icons.
// We simply return a dummy component that renders the icon name.
jest.mock('react-native-vector-icons/Ionicons', () => {
  const {Text} = require('react-native');
  return ({name, size, color, style}) => {
    return (
      // Render the name and other props in a Text element (or string) for testing.
      <Text>{`Ionicon: ${name}`}</Text>
    );
  };
});

// Mock Pile to return a simple component that we can interact with in tests.
// We add a testID using the pieceId so that we can query for it.
jest.mock('../../src/components/Pile', () => {
  return ({onPress, pieceId}) => {
    return (
      <div
        // In react-native testing-library, you can use testID on any element.
        testID={`pile-${pieceId}`}
        // Using onClick here to simulate a press event in our dummy component.
        onClick={onPress}>
        Pile: {pieceId}
      </div>
    );
  };
});

// Mock handleForwardThunk so that it returns a plain object.
// This allows us to compare objects rather than anonymous functions.
jest.mock('../../src/redux/reducers/gameActions', () => ({
  handleForwardThunk: jest.fn((playerNo, pieceId, cellId) => ({
    type: 'HANDLE_FORWARD',
    playerNo,
    pieceId,
    cellId,
  })),
}));

// --- Spy on Redux hooks ---
// We want to control the values returned by useSelector and capture dispatch calls.
const useDispatchMock = jest.spyOn(redux, 'useDispatch');
const useSelectorMock = jest.spyOn(redux, 'useSelector');

describe('Cell Component', () => {
  const dispatchMock = jest.fn();

  beforeEach(() => {
    // Whenever useDispatch is called in Cell, return our mock dispatch.
    useDispatchMock.mockReturnValue(dispatchMock);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders container with safe spot background when the id is in SafeSpots', () => {
    // For this test, we want to check that the container background color
    // is the provided color when the cell id is a safe spot.
    // We assume that SafeSpots is imported from ../helpers/PlotData.
    // For testing purposes, we can temporarily add an id to SafeSpots.
    const safeId = 100;
    // Save the original array so we can restore it after the test.
    const originalSafeSpots = [
      ...require('../../src/helpers/PlotData').SafeSpots,
    ];
    require('../../src/helpers/PlotData').SafeSpots.push(safeId);

    // For this test, we don’t need any pieces, so return an empty array.
    useSelectorMock.mockReturnValue([]);
    // Render the Cell with id equal to safeId.
    const {toJSON} = render(<Cell id={safeId} color="red" />);
    const tree = toJSON();

    // The container should have a background color of "red" because it is a safe spot.
    // (The component uses: backgroundColor: isSafeSpot ? color : Colors.white)
    // You can inspect the tree and assert on the style.
    expect(tree).toMatchSnapshot();

    // Restore SafeSpots array.
    require('../../src/helpers/PlotData').SafeSpots = originalSafeSpots;
  });

  it('renders a star icon when id is in StarSpots', () => {
    // Similar to the safe spot test, we add an id to StarSpots.
    const starId = 200;
    const originalStarSpots = [
      ...require('../../src/helpers/PlotData').StarSpots,
    ];
    require('../../src/helpers/PlotData').StarSpots.push(starId);

    useSelectorMock.mockReturnValue([]);
    render(<Cell id={starId} color="blue" />);

    // Our Ionicons mock renders text starting with "Ionicon:".
    // The Cell component conditionally renders a star icon when isStarSpot is true.
    expect(screen.queryByText('Ionicon: star-outline')).toBeTruthy();

    // Restore StarSpots.
    require('../../src/helpers/PlotData').StarSpots = originalStarSpots;
  });

  it('renders piles for pieces at the given cell and dispatches handleForwardThunk on press', () => {
    // Test scenario: a piece is positioned at this cell.
    const cellId = 5;
    // Dummy piece data that our selector returns. The Cell component filters
    // plotedPieces for items where item.pos == id.
    // For piece id, we use a string that starts with "A" so that playerNo is 1.
    const dummyPieces = [
      {id: 'A1', pos: cellId},
      {id: 'B2', pos: cellId},
      {id: 'C3', pos: cellId},
      {id: 'D4', pos: cellId},
    ];

    useSelectorMock.mockReturnValue(dummyPieces);
    const {getByTestId} = render(<Cell id={cellId} color="green" />);

    // Our mocked Pile should render an element with testID `pile-A1`.
    const pileElement = getByTestId('pile-A1');
    expect(pileElement).toBeTruthy();

    // Simulate a press on the Pile element.
    fireEvent.press(pileElement);

    // Verify that handleForwardThunk was called with the correct arguments.
    expect(handleForwardThunk).toHaveBeenCalledWith(1, 'A1', cellId);

    // Because we mocked handleForwardThunk to return a plain object, we can now compare that object.
    const expectedAction = {
      type: 'HANDLE_FORWARD',
      playerNo: 1,
      pieceId: 'A1',
      cellId,
    };
    expect(dispatchMock).toHaveBeenCalledWith(expectedAction);
  });
});
