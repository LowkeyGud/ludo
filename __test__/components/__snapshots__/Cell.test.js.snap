// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Component renders container with safe spot background when the id is in SafeSpots 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderColor": "grey",
        "borderWidth": 0.5,
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      },
      {
        "backgroundColor": "red",
      },
    ]
  }
/>
`;
